vxportia/router.py

```diff
@@ -0,0 +1,92 @@
+
+from vumi.errors import DispatcherError
+from vumi.dispatchers.base import BaseDispatchRouter
+
+from twisted.internet import reactor
+from twisted.internet.protocol import Factory
+from twisted.internet.endpoints import clientFromString
+from twisted.internet.defer import inlineCallbacks
+
+from vxportia.protocol import PortiaProtocol
+
+
+class PortiaRouter(BaseDispatchRouter):
+    """
+    A router that connects to Portia, the number porting database,
+    to map incoming messages to MNOs and to determine how outbound
+    messages are meant to be routed back.
+    """
+
+    clock = reactor
+
+    @inlineCallbacks
+    def setup_routing(self):
+        self.portia_config = self.config['portia']
+
+        self.mno_transport_map = self.portia_config['mappings']
+        self.transport_mno_map = dict([
+            (transport_name, mno)
+            for mno, transport_name in self.mno_transport_map.items()])
+        self.default_transport = self.portia_config.get('default_transport')
+
+        known_transport_names = self.mno_transport_map.values()
+        if self.default_transport:
+            known_transport_names.append(self.default_transport)
+
+        if (set(known_transport_names) !=
+                set(self.config['transport_names'])):
+            raise DispatcherError(
+                'PortiaRouter mappings differ from dispatcher '
+                'transport_names.')
+
+        if len(self.config['exposed_names']) != 1:
+            raise DispatcherError(
+                ('PortiaRouter is only able to work with 1 exposed_name, '
+                 'there are %s configured.') % (
+                    len(self.config['exposed_names'],)))
+
+        self.exposed_name = self.config['exposed_names'][0]
+
+        self.endpoint = clientFromString(
+            self.clock, self.portia_config['client_endpoint'])
+        self.portia = yield self.endpoint.connect(
+            Factory.forProtocol(PortiaProtocol))
```

Comment:
Please don't drag me into Twisted's reconnecting endpoint swamp ... Ideally I would love for these things to take Erlang's approach, have it crash and burn and have some supervisor restart it and carry on.